internal/agent/nftables_generator.go

```diff
@@ -73,19 +141,156 @@ type nftablesConfig struct {
 	FWMarkBits              uint32
 	FWMarkMask              uint32
 	Forwards                []nftablesForward
+	NetworkPolicies         map[string]networkPolicy
+	PolicyAssignments       []policyAssignment
 }
 
 type NftablesGenerator struct {
 	Cfg config.Nftables
 }
 
+// Takes a slice of strings and produces a list ready to be used in an nftables rule
+// eg. []string{"1", "2", "3"} => "{1,2,3}"
+func makeNftablesList(items []string) string {
+	return "{" + strings.Join(items, ",") + "}"
+}
+
+//TODO: func to check if a string is an ip address or cidr. validate!
+
+func makePortMatches(in []model.PolicyPort) (portMatches []string) {
+	portMap, err := makePortMap(in)
+	if err != nil {
+		return
+	}
+	portMatches = make([]string, 0, len(portMap))
+	for proto, ports := range portMap {
+		newMatch := proto
+		if len(ports) != 0 {
+			newMatch += " dport " + makeNftablesList(ports)
+		}
+		portMatches = append(portMatches, newMatch)
+	}
+	// if there are no port matches, all ports are allowed
+	if len(portMatches) == 0 {
+		portMatches = append(portMatches, "")
+	}
+	return portMatches
+}
+
+func makeSAddrMatches(in []model.AllowedIPBlock) (SAddrMatches []SAddrMatch) {
+	if len(in) == 0 {
+		return []SAddrMatch{SAddrMatch{}}
+	}
```

Comment:
I agree that the behavior is not very obvious from the code and I would like it to be more explicit as well. The problem I have with injecting 0.0.0.0/0 and ::/0 is the resulting inconsistency between `from` and `ports`.

With the current design we have:
* No `from` => All IP addresses allowed
* No `ports` => All ports allowed

Were we to inject 0.0.0.0/0 and ::/0, we'd need to default block IP ranges which would result in:
* No `from` => All IP addresses blocked
* No `ports` => All ports allowed

IMO this can be just as confusing when implementing a potential future agent.

Reply:
The first design iteration had a `defaultPolicy` element in the struct for this purpose which could be set to 'block' or 'allow' be the controller. I'm thinking about bringing it back...

Reply:
Riiight, the ports thing.... There it's also more complicated to specify a "wildcard" to inject by default, because no port filtering also implies no protocol filtering (e.g. ICMP, which cannot be represented as a port number).

I'm not so keen on a "defaultPolicy", because it would only ever be set to "allow" if the AllowedIPBlocks array is empty, and set to "block" otherwise. That is weird, and I'd prefer signalling via the empty array instead.

That said, is there maybe a better name for `AllowedIPBlocks` we could use? `IPBlockFilters`, and then rename the CIDR field to `Allow` and the `Except` field to `Block`?

Then the behaviour of an empty list becomes less obvious ("nothing in allowed" is normally "obviously" "everything blocked") and that means people are more likely to consult documentation where we can explain the misery we're in.

Reply:
I like that! How about:
```go
type IPBlockFilter struct {
	Allow string   `json:"allow"`
	Block []string `json:"block"`
}

type PortFilter struct {
	Protocol corev1.Protocol `json:"protocol"`
	Port     *int32          `json:"port,omitempty"` // Don't filter by port number if empty (only by protocol)
	EndPort  *int32          `json:"end-port,omitempty"`
}

type AllowedIngress struct {
	IPBlockFilters []IPBlockFilter `json:"ipblock-filter"` // Don't filter by address if empty (allow all)
	PortFilters    []PortFilter    `json:"port-filter"`    // Don't filter by transport protocol or port if empty (allow all)
}

type NetworkPolicy struct {
	Name             string           `json:"name"`
	AllowedIngresses []AllowedIngress `json:"allowed-ingresses"` // Block all incoming traffic if empty
}
```

The name `AllowedIngresses` (although it's somewhat ugly) implies that ingress should be blocked if it's empty.
Having an `AllowedIngress` struct without filter implies that nothing should be filtered, hence all ingress is allowed.

Reply:
I think golang has documentation comment conventions which you should follow, otherwise LGTM.